openssl/src/crypto/sign.rs

```diff
@@ -0,0 +1,274 @@
+use ffi;
+use std::io::{self, Write};
+use std::marker::PhantomData;
+use std::ptr;
+
+use HashTypeInternals;
+use crypto::hash::Type;
+use crypto::pkey::PKey;
+use error::ErrorStack;
+
+#[cfg(ossl110)]
+use ffi::{EVP_MD_CTX_new, EVP_MD_CTX_free};
+#[cfg(any(ossl101, ossl102))]
+use ffi::{EVP_MD_CTX_create as EVP_MD_CTX_new, EVP_MD_CTX_destroy as EVP_MD_CTX_free};
+
+pub struct Signer<'a>(*mut ffi::EVP_MD_CTX, PhantomData<&'a PKey>);
+
+impl<'a> Drop for Signer<'a> {
+    fn drop(&mut self) {
+        unsafe {
+            EVP_MD_CTX_free(self.0);
+        }
+    }
+}
+
+impl<'a> Signer<'a> {
+    pub fn new(type_: Type, pkey: &'a PKey) -> Result<Signer<'a>, ErrorStack> {
+        unsafe {
+            ffi::init();
+
+            let ctx = try_ssl_null!(EVP_MD_CTX_new());
+            let r = ffi::EVP_DigestSignInit(ctx,
+                                            ptr::null_mut(),
+                                            type_.evp_md(),
+                                            ptr::null_mut(),
+                                            pkey.as_ptr());
+            if r != 1 {
+                EVP_MD_CTX_free(ctx);
+                return Err(ErrorStack::get());
+            }
+
+            Ok(Signer(ctx, PhantomData))
+        }
+    }
+
+    pub fn update(&mut self, buf: &[u8]) -> Result<(), ErrorStack> {
+        unsafe {
+            try_ssl_if!(ffi::EVP_DigestUpdate(self.0, buf.as_ptr() as *const _, buf.len()) != 1);
+            Ok(())
+        }
+    }
+
+    pub fn finish(&self) -> Result<Vec<u8>, ErrorStack> {
+        unsafe {
+            let mut len = 0;
+            try_ssl_if!(ffi::EVP_DigestSignFinal(self.0, ptr::null_mut(), &mut len) != 1);
+            let mut buf = vec![0; len];
+            try_ssl_if!(ffi::EVP_DigestSignFinal(self.0, buf.as_mut_ptr() as *mut _, &mut len)
+                    != 1);
+            Ok(buf)
+        }
+    }
+}
+
+impl<'a> Write for Signer<'a> {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
```

Comment:
As discussed in #420 this (and `update`) would have problems if a write of a buffer greater than `u32::max_value()` bytes in size are attempted.  The [ostrich algorithm](https://en.wikipedia.org/wiki/Ostrich_algorithm) might be fine for this case, but making a comment as it is one difference with the implementation from the previous PR.